packages/vite/src/node/server/mixedModuleGraph.ts

```diff
@@ -452,8 +458,36 @@ export class ModuleGraph {
     clientModule?: EnvironmentModuleNode,
     ssrModule?: EnvironmentModuleNode,
   ): ModuleNode {
-    // ...
-    return new ModuleNode(this, clientModule, ssrModule)
+    const cached = this.moduleNodeCache.get(clientModule, ssrModule)
+    if (cached) {
+      return cached
+    }
+
+    const moduleNode = new ModuleNode(this, clientModule, ssrModule)
+    this.moduleNodeCache.set(clientModule, ssrModule, moduleNode)
+    return moduleNode
+  }
+}
+
+class DualWeakMap<K1 extends WeakKey, K2 extends WeakKey, V> {
+  private map = new WeakMap<K1 | object, WeakMap<K2 | object, V>>()
+  private undefinedKey = {}
+
+  get(key1: K1 | undefined, key2: K2 | undefined): V | undefined {
+    const k1 = key1 ?? this.undefinedKey
+    const k2 = key2 ?? this.undefinedKey
+    return this.map.get(k1)?.get(k2)
+  }
+
+  set(key1: K1 | undefined, key2: K2 | undefined, value: V): void {
+    const k1 = key1 ?? this.undefinedKey
+    const k2 = key2 ?? this.undefinedKey
+    if (!this.map.has(k1)) {
+      this.map.set(k1, new Map<K2, V>())
+    }
+
+    const m = this.map.get(k1)!
+    m.set(k2, value)
   }
```

Comment:
Due to the simplicity of this implementation, a memory leak occurs in one particular case.
```js
const dualWeakMap = new DualWeakMap()
const k1 = {}
const k2 = {}
dualWeakMap.set(k1, k2)
//
// some time passes and k2 gets dropped from memory, but k1 was kept
//

// this should return false because there's no entries in `dualWeakMap.map.get(k1)`
// but this return true
// since this implementation does not remove the inner map when the entry in the inner map is dropped
dualWeakMap.map.has(k1)
```
But I think this is fine for our case. Modules doesn't get dropped usually, and I assume the memory size of an empty Map is not that big.

Reply:
Yeah I think it should be fine too. If k2 (which refers to the `ssrModule`) gets dropped and k1 (the `clientModule`) still exists, then I think it'll at least be cached with the `clientModule` + `undefinedKey` pair so the weakmap is still useful.

The tricky part I guess is that `undefinedKey` would always exist which would cause the memory leak.